checkers/testdata/deferUnlambda/negative_tests.go

```diff
@@ -43,6 +43,17 @@ func negativeTests() {
 	defer func() {
 		objects[0].f()
 	}()
+
+	// OK: func foo is reassigned later #1401
+	foo := func() {
+		println("should never print")
+	}
+	defer func() {
+		foo()
+	}()
+	foo = func() {
+		println("should print")
+	}
```

Comment:
Based on the implementation, I guess it wouldn't "see" the assignment anywhere except the same depth level.
```go
// This assignment inside of the if statement will not be handled.
if cond {
	foo = func() {
		println("should print")
	}
}
```
The proper analysis is cumbersome at AST level. This is probably one of the cases where SSA would be beneficial.

Reply:
Thank you for your advice. Since this is the first time I work with AST, I'll see what I can do in my time. Meanwhile, I mark this PR as a draft until I figure out a proper solution based on what you mentioned.